src/target/mspm0.c

```diff
@@ -0,0 +1,349 @@
+#define MSPM0_FLASH_FACTORY   0x41c40000U /* One Sector, BANK0. Non modifiable */
+#define MSPM0_FLASH_DATA      0x41d00000U
+#define MSPM0_FLASH_SECTOR_SZ 1024
+
+#define MSPM0_FLASHCTL_BASE              0x400cd000
+#define MSPM0_FLASHCTL_CMDEXEC           (MSPM0_FLASHCTL_BASE + 0x1100)
+#define MSPM0_FLASHCTL_CMDTYPE           (MSPM0_FLASHCTL_BASE + 0x1104)
+#define MSPM0_FLASHCTL_CMDCTL            (MSPM0_FLASHCTL_BASE + 0x1108)
+#define MSPM0_FLASHCTL_CMDADDR           (MSPM0_FLASHCTL_BASE + 0x1120)
+#define MSPM0_FLASHCTL_BYTEN             (MSPM0_FLASHCTL_BASE + 0x1124)
+#define MSPM0_FLASHCTL_CMDDATAIDX        (MSPM0_FLASHCTL_BASE + 0x112c)
+#define MSPM0_FLASHCTL_STATCMD           (MSPM0_FLASHCTL_BASE + 0x13d0)
+#define MSPM0_FLASHCTL_CMDDATA0          (MSPM0_FLASHCTL_BASE + 0x1130)
+#define MSPM0_FLASHCTL_CMDDATA1          (MSPM0_FLASHCTL_BASE + 0x1134)
+#define MSPM0_FLASHCTL_CMDDATA2          (MSPM0_FLASHCTL_BASE + 0x1138)
+#define MSPM0_FLASHCTL_CMDDATA3          (MSPM0_FLASHCTL_BASE + 0x113c)
+#define MSPM0_FLASHCTL_CMDWEPROTA        (MSPM0_FLASHCTL_BASE + 0x11d0)
+#define MSPM0_FLASHCTL_CMDWEPROTB        (MSPM0_FLASHCTL_BASE + 0x11d4)
+#define MSPM0_FLASHCTL_CMDWEPROTC        (MSPM0_FLASHCTL_BASE + 0x11d8)
+#define MSPM0_FLASHCTL_CMDWEPROTNM       (MSPM0_FLASHCTL_BASE + 0x1210)
+#define MSPM0_FLASHCTL_CMDTYPE_NOOP      0U
+#define MSPM0_FLASHCTL_CMDTYPE_PROG      1U
+#define MSPM0_FLASHCTL_CMDTYPE_ERASE     2U
+#define MSPM0_FLASHCTL_CMDTYPE_RDVERIFY  3U
+#define MSPM0_FLASHCTL_CMDTYPE_BLVERIFY  6U
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_1WORD  (0U << 4)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_2WORDS (1U << 4)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_4WORDS (2U << 4)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_8WORDS (3U << 4)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_SECTOR (4U << 4)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_BANK   (5U << 4)
+#define MSPM0_FLASHCTL_CMDEXEC_EXEC      1U
+#define MSPM0_FLASHCTL_STAT_DONE         0x01U
+#define MSPM0_FLASHCTL_STAT_CMDPASS      0x02U
+
+typedef struct mspm0_flash {
+	target_flash_s target_flash;
+	uint32_t banks;
+} mspm0_flash_s;
+
+static bool mspm0_flash_erase(target_flash_s *flash, target_addr_t addr, size_t length);
+static bool mspm0_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t length);
+static bool mspm0_mass_erase(target_s *target);
+
+#if MSPM0_CONFIG_FLASH_DUMP_SUPPORT
+static bool mspm0_dump_factory_config(target_s *const target, const int argc, const char **const argv);
+static bool mspm0_dump_bcr_config(target_s *const target, const int argc, const char **const argv);
+
+static command_s mspm0_cmds_list[] = {{"dump_factory", mspm0_dump_factory_config, "Display FACTORY registers"},
+	{"dump_bcr", mspm0_dump_bcr_config, "Display NONMAIN (BCR/BSL) registers"},
+	/* {"write_nm", mspm0_write_non_main, "Write data to NONMAIN (configuration) flash. write_nm <address> <value>"}, */
+	{NULL, NULL, NULL}};
+
+typedef struct conf_register {
+	uint16_t offset : 12;
+	uint16_t words : 4;
```

Comment:
Please do not use bitfields. They are not well defined by the C specification and so their use is badly liable to causing undefined behaviour and unwanted side-effects, up to and including this structure winding up padded to `alignof(int)` instead of `alignof(uint16_t)`, negating the entire purpose of this.

This is why the code base prefers masking and shifting and an encoding macro instead. Much easier to control and get the behaviour desired.